pkgs/web_socket/lib/src/web_socket.dart

```diff
@@ -0,0 +1,135 @@
+  final int? code;
+
+  /// A textual explanation of the reason why the WebSocket was closed.
+  ///
+  /// Will be empty if the peer did not specify a reason.
+  final String reason;
+
+  CloseReceived([this.code, this.reason = '']);
+
+  @override
+  bool operator ==(Object other) =>
+      other is CloseReceived && other.code == code && other.reason == reason;
+
+  @override
+  int get hashCode => [code, reason].hashCode;
+
+  @override
+  String toString() => 'CloseReceived($code, $reason)';
+}
+
+class WebSocketException implements Exception {
+  final String message;
+  WebSocketException([this.message = '']);
+}
+
+/// Thrown if [WebSocket.sendText], [WebSocket.sendBytes], or
+/// [WebSocket.close] is called when the [WebSocket] is closed.
+class WebSocketConnectionClosed extends WebSocketException {
+  WebSocketConnectionClosed([super.message = 'Connection Closed']);
+}
+
+/// The interface for WebSocket connections.
+///
+/// TODO: insert a usage example.
+abstract interface class WebSocket {
+  /// Sends text data to the connected peer.
+  ///
+  /// Throws [WebSocketConnectionClosed] if the [WebSocket] is
+  /// closed (either through [close] or by the peer).
+  void sendText(String s);
+
+  /// Sends binary data to the connected peer.
+  ///
+  /// Throws [WebSocketConnectionClosed] if the [WebSocket] is
+  /// closed (either through [close] or by the peer).
+  void sendBytes(Uint8List b);
+
+  /// Closes the WebSocket connection and the [events] `Stream`.
+  ///
+  /// Sends a Close frame to the peer. If the optional [code] and [reason]
+  /// arguments are given, they will be included in the Close frame. If no
+  /// [code] is set then the peer will see a 1005 status code. If no [reason]
+  /// is set then the peer will not receive a reason string.
+  ///
+  /// Throws a [RangeError] if [code] is not in the range 3000-4999.
+  ///
+  /// Throws an [ArgumentError] if [reason] is longer than 123 bytes when
+  /// encoded as UTF-8
+  ///
+  /// Throws [WebSocketConnectionClosed] if the connection is already
+  /// closed (including by the peer).
+  Future<void> close([int? code, String? reason]);
+
+  /// A [Stream] of [WebSocketEvent] received from the peer.
+  ///
+  /// Data received by the peer will be delivered as a [TextDataReceived] or
+  /// [BinaryDataReceived].
+  ///
+  /// If a [CloseReceived] event is received then the [Stream] will be closed. A
+  /// [CloseReceived] event indicates either that:
+  ///
+  /// - A close frame was received from the peer. [CloseReceived.code] and
+  ///   [CloseReceived.reason] will be set by the peer.
+  /// - A failure occured (e.g. the peer disconnected). [CloseReceived.code] and
+  ///   [CloseReceived.reason] will be a failure code defined by
+  ///   (RFC-6455)[https://www.rfc-editor.org/rfc/rfc6455.html#section-7.4.1]
+  ///   (e.g. 1006).
+  ///
+  /// Errors will never appear in this [Stream].
+  Stream<WebSocketEvent> get events;
```

Comment:
This Stream and its relationship with the `sendText`/`sendBytes`/`close` methods is my biggest concern with this API.

As documented now, `sendText`, etc. throw if the WebSocket is closed. But the `ClosedReceived` event is delivered asynchronously so you can't rely on it being delivered in time to prevent those calls. For example:

```dart
timer = Timer.periodic(const Duration(seconds: 2), (timer) {
  websocket.sendText('<stock>GOOG</stock>');
});

await for (final event in websocket.events) {
  switch (event) {
  // Should probably do something with the stock quote.
  case CloseReceived:
    // This is not sufficient - the implementation may add `CloseReceived` to the `StreamController` and then
    // close the `StreamController` before this event is received. This probably means that, in practice, every
    // call to `sendText`/`sendBytes` will need to be guarded in a `try`/`catch`
    timer.cancel();
  }
}
```

Some alternatives:
1. just throw away `sendText`/`sendBytes` after the WebSocket is closed - this is what JavaScript does
2. use `SyncStreamController` (will that work?)
3. make a synchronous callback for events instead of using a `Stream`.

Reply:
I think even if we synchronously surface a `ClosedReceived` there is still the possibility of a `sendText` failing to write because of a closed socket, though it may be significantly less less likely.

I would probably start with throwing, and we can consider an option to silently discard if that has a clear use case.

Reply:
> I think even if we synchronously surface a ClosedReceived there is still the possibility of a sendText failing to write because of a closed socket, though it may be significantly less less likely.

My plan was for `sendText`/`sendBytes` to not throw in that case. Any other behavior would be difficult on the web, [which only allows `send` to throw if the connection isn't established](https://websockets.spec.whatwg.org/#dom-websocket-send).

I have a [conformance test to verify that behavior](https://github.com/dart-lang/http/blob/7b30f614cfda972960f0a48017587a64d61c1781/pkgs/web_socket_conformance_tests/lib/src/disconnect_after_upgrade_tests.dart) (the tests are still very rough at this point).

I've updated the docs to make a note of that.

> I would probably start with throwing, and we can consider an option to silently discard if that has a clear use case.

SGTM!